src/saml.ts

```diff
@@ -462,90 +373,50 @@ class SAML {
   }
 
   async _requestToUrlAsync(
-    request: string | null | undefined,
+    request: string | null,
     response: string | null,
     operation: string,
     additionalParameters: querystring.ParsedUrlQuery
   ): Promise<string> {
     this.options.entryPoint = assertRequired(this.options.entryPoint, "entryPoint is required");
 
-    let buffer: Buffer;
-    if (this.options.skipRequestCompression) {
-      buffer = Buffer.from((request || response)!, "utf8");
-    } else {
-      buffer = await deflateRawAsync((request || response)!);
-    }
+    const targetUrl =
+      operation === "logout" && this.options.logoutUrl
+        ? this.options.logoutUrl
+        : this.options.entryPoint;
 
-    const base64 = buffer.toString("base64");
-    let target = new URL(this.options.entryPoint);
+    const message = request || response;
+    const messageType = request ? "SAMLRequest" : "SAMLResponse";
 
-    if (operation === "logout") {
-      if (this.options.logoutUrl) {
-        target = new URL(this.options.logoutUrl);
-      }
-    } else if (operation !== "authorize") {
-      throw new Error("Unknown operation: " + operation);
+    if (!message) {
+      throw new Error("response or request should be provided");
     }
 
-    const samlMessage: querystring.ParsedUrlQuery = request
-      ? {
-          SAMLRequest: base64,
-        }
-      : {
-          SAMLResponse: base64,
-        };
-    Object.keys(additionalParameters).forEach((k) => {
-      samlMessage[k] = additionalParameters[k];
-    });
-    if (isValidSamlSigningOptions(this.options)) {
-      if (!this.options.entryPoint) {
-        throw new Error('"entryPoint" config parameter is required for signed messages');
-      }
-
-      // sets .SigAlg and .Signature
-      this.signRequest(samlMessage);
-    }
-    Object.keys(samlMessage).forEach((k) => {
-      target.searchParams.set(k, samlMessage[k] as string);
+    return await requestToUrl({
+      targetUrl,
+      skipRequestCompression: this.options.skipRequestCompression,
+      message,
+      messageType,
+      additionalParameters,
+      privateKey: this.options.privateKey,
+      signatureAlgorithm: this.options.signatureAlgorithm,
     });
-
-    return target.toString();
   }
 
   _getAdditionalParams(
```

Comment:
Ideally, this wouldn't need to be here at all. Is there a reason it wasn't completely factored out?

Reply:
Sure, let's be safer for this massive PR.